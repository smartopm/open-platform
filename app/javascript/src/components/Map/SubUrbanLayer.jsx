import React, { useEffect } from 'react'
import { useLeaflet } from 'react-leaflet'
import L from 'leaflet'
import NkwashiSuburbBoundaryData from '../../data/nkwashi_suburb_boundary.json'
import SubUrbanLegendContent from './SubUrbanLegendContent'
import { getHexColor } from '../../utils/helpers'

function geoJSONStyle(feature) {
  return {
    weight: 2,
    opacity: 1,
    color: "white",
    dashArray: "3",
    fillOpacity: 0.7,
    fillColor: getHexColor(feature && Number(feature.properties.estimated_population))
  }
}

export default function SubUrbanLayer(){
  const { map } = useLeaflet()

  /* istanbul ignore next */
  useEffect(() => {
    // Used to manage control on the map
    const info = L.control()

    info.onAdd = () => {
      info.div = L.DomUtil.create('div', 'info')
      info.update()
      return info.div
    };

    /* istanbul ignore next */
    /* eslint-disable react/prop-types */
    info.update = function(props){
      info.div.innerHTML =
        `<h5>Est. Population Census/Nkwashi Sub-urban</h5>
        <p>Index: 1 Plot == 1 Building == 1 Household </p>${ 
        props
          ? `<b>${ 
            props.sub_urban 
            }</b><br />${ 
            Number(props.estimated_population)
            } people / mi<sup>2</sup>`
          : 'Hover over a Sub-urban area'}`
    };

    info.addTo(map)

    /* istanbul ignore next */
    const highlightFeature = function(e) {
      const layer = e.target;

      layer.setStyle({
        weight: 5,
        color: '#666',
        dashArray: '',
        fillOpacity: 0.7
      })

      info.update(layer.feature.properties)
    };

    let geojson

    /* istanbul ignore next */
    const resetHighlight = function(e) {
      geojson.resetStyle(e.target)
      info.update()
    };

    /* istanbul ignore next */
    const zoomToFeature = function(e) {
      map.fitBounds(e.target.getBounds())
    };

    /* istanbul ignore next */
    const onEachFeature = (feature, layer) => {
      layer.on({
        mouseover: highlightFeature,
        mouseout: resetHighlight,
        click: zoomToFeature
      })
    }

    geojson = L.geoJson(NkwashiSuburbBoundaryData, {
      style: geoJSONStyle,
      onEachFeature
    })?.addTo(map)
  }, [map])

  return (
    <>
      <style
         // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{
          __html: `
        .info {
        padding: 6px 8px;
        font: 14px/16px Arial, Helvetica, sans-serif;
        background: white;
        background: rgba(255, 255, 255, 0.8);
        box-shadow: 0 0 15px rgba(0, 0, 0, 0.2);
        border-radius: 5px;
        margin-left: 70px;
      }

      .info h5 {
        margin: 0 0 5px;
        color: #777;
      }

      .legend {
        text-align: left;
        line-height: 18px;
        color: #555;
      }

      .legend i {
        width: 18px;
        height: 18px;
        float: left;
        margin-right: 8px;
        opacity: 0.7;
      }
      `
        }}
      />
      <SubUrbanLegendContent />
    </>
    )
}