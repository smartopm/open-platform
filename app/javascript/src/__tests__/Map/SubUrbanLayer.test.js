import React from 'react'
import { MockedProvider } from '@apollo/react-testing'
import { BrowserRouter } from 'react-router-dom'
import { act, render } from '@testing-library/react'
import SubUrbanLayer from '../../components/Map/SubUrbanLayer'

describe('<SubUrbanLayer />', () => {
  it('should render SubUrbanLayer', async () => {
    let container;
    await act(async () => {
      container = render(
        <MockedProvider>
          <BrowserRouter>
            <SubUrbanLayer />
          </BrowserRouter>
        </MockedProvider>
      )
    })
    expect(container).toMatchSnapshot()
  })
})
