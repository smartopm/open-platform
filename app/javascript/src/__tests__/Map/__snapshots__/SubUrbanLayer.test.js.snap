// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubUrbanLayer /> should render SubUrbanLayer 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <style>
        
        .info {
        padding: 6px 8px;
        font: 14px/16px Arial, Helvetica, sans-serif;
        background: white;
        background: rgba(255, 255, 255, 0.8);
        box-shadow: 0 0 15px rgba(0, 0, 0, 0.2);
        border-radius: 5px;
      }

      .info h5 {
        margin: 0 0 5px;
        color: #777;
      }

      .legend {
        text-align: left;
        line-height: 18px;
        color: #555;
      }

      .legend i {
        width: 18px;
        height: 18px;
        float: left;
        margin-right: 8px;
        opacity: 0.7;
      }
      
      </style>
    </div>
  </body>,
  "container": <div>
    <style>
      
        .info {
        padding: 6px 8px;
        font: 14px/16px Arial, Helvetica, sans-serif;
        background: white;
        background: rgba(255, 255, 255, 0.8);
        box-shadow: 0 0 15px rgba(0, 0, 0, 0.2);
        border-radius: 5px;
      }

      .info h5 {
        margin: 0 0 5px;
        color: #777;
      }

      .legend {
        text-align: left;
        line-height: 18px;
        color: #555;
      }

      .legend i {
        width: 18px;
        height: 18px;
        float: left;
        margin-right: 8px;
        opacity: 0.7;
      }
      
    </style>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
